src/ui/public/filter_bar/lib/multi_filter.js

```diff
@@ -0,0 +1,93 @@
+import _ from 'lodash';
+
+/**
+ * Create single filter per data field. Fields with multiple filters are merged into a bool query.
+ * Filters without meta.key can not be merged because field is unknown.
+ *
+ * @param {array} filters to merge.
+ * @returns {array} merged filters
+ */
+export function mergeFilters(filters) {
+  // Group filters by key
+  const filterGroups = new Map();
+  const filtersWithNoKey = [];
+  filters.forEach((filter) => {
+    const key = _.get(filter, 'meta.key');
+    if (!key) {
+      filtersWithNoKey.push(filter);
+      return;
+    }
+
+    if (!filterGroups.has(key)) {
+      filterGroups.set(key, [filter]);
+    } else {
+      filterGroups.set(key, filterGroups.get(key).concat(filter));
+    }
+  });
```

Comment:
You could probably replace this `forEach` with a `_.groupBy(filters, 'meta.key')`